ServeStaticFilesFrom with separate route and URL prefixes. URL redirects refactor.

sompylasar opened the pull request:
The issue I'm solving here is issuing a redirect from the backend with the URL understood by the end user, even if the URL is transformed along the way from the server to the user (e.g. on a proxy).

Unfortunately, this required the changes to the `URL` class behavior: its "previous URL" handling and defaults handling. There were too many variants of use of this class's options, some of which were conflicting with each other (e.g. default scheme with an unknown port or no scheme with an unknown port).

The `HTTPClient` was also changed to throw if the redirects weren't allowed before following any redirect response code.

The `SocketResolveAddressException` was filled with the resolve arguments to have more details about the error in runtime.

CC @dkorolev @grixa @mzhurovich

dkorolev commented on Blocks/HTTP/test.cc:
```diff
@@ -357,6 +357,28 @@ TEST(HTTPAPI, Redirect) {
+TEST(HTTPAPI, RedirectToFullURL) {
+  // TODO(sompylasar): Cannot test redirect to a different port because the HTTP client
+  // is following the redirect and tries to connect to the redirect target,
+  // and we do not have a second port to occupy for the tests.
```
You can totally use a second port if you want to.

sompylasar replied:
Should I add a "flag" for it then, or just this port plus 1 will suffice?

dkorolev replied:
A flag would be better. It'll allocate a new port, yes.

sompylasar replied:
Done. https://github.com/C5T/Current/pull/744/commits/c795aa3915da1446ce3d574a13bc35efa55420fb

dkorolev commented:
Looks like the

```
[ RUN      ] HTTPAPI.ServeStaticFilesFromOptionsCustomPublicRoutePrefixAndPublicUrlPrefix
unknown file: Failure
C++ exception with description "docu/client/../../../../Bricks/net/http/impl/../../tcp/impl/posix.h:557	SocketConnectException()	" thrown in the test body.
[  FAILED  ] HTTPAPI.ServeStaticFilesFromOptionsCustomPublicRoutePrefixAndPublicUrlPrefix (3 ms)
```

test is consistently failing on Travis.

sompylasar commented:
RE: SocketConnectException, weird, will look into it.

mzhurovich commented on Blocks/URL/url.h:
```diff
@@ -356,7 +359,7 @@ struct URL : URLParametersExtractor, URLWithoutParametersParser {
-      const std::string& previous_scheme = kDefaultScheme,
+      const std::string& previous_scheme = "",
       const std::string& previous_host = "",
       const int previous_port = 0)
```
We've already changed `int port` to `uint16_t` port in some places - let's fix it here as well?

sompylasar replied:
Sure.

sompylasar replied:
Done. https://github.com/C5T/Current/pull/744/commits/c795aa3915da1446ce3d574a13bc35efa55420fb

mzhurovich commented on Blocks/HTTP/impl/posix_server.h:
```diff
@@ -126,7 +141,7 @@ struct StaticFileServer {
         // See RFC1808, `Resolving Relative URLs`, `Step 6`: https://www.ietf.org/rfc/rfc1808.txt
         r.connection.SendHTTPResponse(
-            "", HTTPResponseCode.Found, content_type, current::net::http::Headers({{"Location", r.url.path + '/'}}));
+            "", HTTPResponseCode.Found, content_type, current::net::http::Headers({{"Location", trailing_slash_redirect_url}}));
```
Should the Location header be set if `trailing_slash_redirect_url` is empty?

sompylasar replied:
Good question, I assumed this class to be of internal use, and in this case it won't be empty, but yes, I can add a check to be safe.

sompylasar replied:
Done. https://github.com/sompylasar/Current/commit/1c658d0f6b265ef1f79728c59ffa4a223e98b1bd

mzhurovich commented on Blocks/HTTP/test.cc:
```diff
@@ -1068,13 +1095,60 @@ TEST(HTTPAPI, ServeStaticFilesFromOptionsCustomURLBase) {
+               HTTPRedirectNotAllowedException);
+  const auto sub_sub_dir_response = HTTP(GET(Printf("http://localhost:%d/static/something/sub_dir/sub_sub_dir", FLAGS_net_api_test_port)).AllowRedirects());
+  EXPECT_EQ("http://localhost/anything/sub_dir/sub_sub_dir/", sub_sub_dir_response.url);
+}
```
Running `make indent` won't hurt

mzhurovich commented on Bricks/net/tcp/impl/posix.h:
```diff
@@ -517,10 +517,10 @@ inline addrinfo_t GetAddrInfo(const std::string& host, const std::string& serv =
   const int retval = ::getaddrinfo(host.c_str(), serv.c_str(), &hints, &result);
   if (!result) {
-    CURRENT_THROW(SocketResolveAddressException());
+    CURRENT_THROW(SocketResolveAddressException(host + ' ' + serv));
```
👍

mzhurovich commented on Blocks/HTTP/impl/posix_server.h:
```diff
@@ -88,33 +88,48 @@ struct ServeStaticFilesFromCannotServeMoreThanOneIndexFile : ServeStaticFilesExc
+  std::string public_route_prefix;
+
+  // User-facing URL prefix. Used to prefix the route in the trailing slash redirects.
+  std::string public_url_prefix;
```
Isn't the route above a "user-facing" as well?

sompylasar replied:
Not in every case, that's why this PR. Imagine: user-facing URL is `http://api.example.com/files/*` but the server serves at `/static/*`. If the server has to issue a trailing slash redirect for `http://api.example.com/files/somedir`, it should return either `/files/somedir/` or `http://api.example.com/files/somedir/`, neither of these contain `/static`

mzhurovich commented on Blocks/URL/test.cc:
```diff
@@ -49,7 +49,7 @@ TEST(URLTest, SmokeTest) {
   EXPECT_EQ("www.google.com", u.host);
   EXPECT_EQ("/", u.path);
-  EXPECT_EQ("http", u.scheme);
+  EXPECT_EQ("", u.scheme);  // Since there is no rule for port "8080", no scheme is returned.
```
👍

mzhurovich commented on Blocks/URL/test.cc:
```diff
@@ -95,42 +101,58 @@ TEST(URLTest, CompositionTest) {
+  EXPECT_EQ("https://www.google.com/", URL("https://www.google.com:443").ComposeURL());
+  EXPECT_EQ("https://www.google.com:444/", URL("https://www.google.com:444").ComposeURL());
+  // Since there is no rule for port "23", no scheme is returned.
+  EXPECT_EQ("www.google.com:23/", URL("www.google.com:23").ComposeURL());
```
In its previous version, this test was explicitly testing a behavior with URL ctor parameters dealing with redirects. I'd suggest to keep the separate tests for port-scheme mapping and tests for derived scheme.

sompylasar replied:
Agree. This was about morning, I could barely wrap my head around the obscure logic of the "previous scheme", "previous port", "previous host", their default values and implicit behavior, etc. I'd rather pull the redirect logic out of `URL` constructor if that sounds better.

sompylasar commented:
RE: `SocketConnectException` in `HTTPAPI.ServeStaticFilesFromOptionsCustomPublicRoutePrefixAndPublicUrlPrefix`, this is because it likely goes to port `80` somewhere; it started failing when I turned off the local `nginx` which was serving port `80` on my machine; so please be aware of that implicit dependency.

Sometimes, especially in low-level APIs, default values cause troubles.

sompylasar commented:
And obviously, the public URL where the redirects go is configured without a port, so `80` is assumed. `"http://localhost/anything"});`
This test needs the secondary port server.

sompylasar commented:
This fix https://github.com/C5T/Current/pull/744/commits/4f397cb9f34c0589d3f0b2fc7304b0232f19922d caused `HTTPRedirectLoopException` 😞  investigating, hold on...

dkorolev commented:
@sompylasar, in lieu of best TDD practices, could you show the shortest possible test that fails w/ the current code to illustrate the issue you're working on fixing?

I still struggle to get to the root of it. Thanks!

sompylasar commented:
@dkorolev I'm sorry, I got used to push immediately to save the work in the cloud/central server, and run the tests in parallel to that. But pushing to a PR incurs extra work: both human (review) and CI (build), and I keep forgetting that. I've fixed that, added assertions and extended exception info, see latest commits.

sompylasar commented:
I am now reworking the `URL` module to have an explicit `RedirectToURL` method which deals with all the redirect-related defaults and substitutions. The logic is too complex and spaghetti to fit into the single constructor.

dkorolev commented:
Oh, so it's the same as Hypermedia base for our REST/CRUD storage. Got it, thanks!

Sure thing, OK to make small commits. My TDD commit was not re. those, but re. it being unclear to me before your last commit what the root issue is.

sompylasar commented:
`URL` class reworked, see: https://github.com/C5T/Current/pull/745 (merged here as well).

sompylasar commented:
Closing in favor of https://github.com/C5T/Current/pull/746
